### ISSUE TITLE
Should the model's grads be cleared before entering the validation loop?

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/lightning#18713]
### Discussion

Our trainer performs optimization in this order: 1) `zero_grad()` 2) `backward()` 3) `optimizer.step()`
This means if validation is run after such an optimization step (either the last one in the epoch or according to the Trainer's `val_check_interval` during the epoch), gradients would still be around. 
To minimize the memory usage and risk of OOM during validation, we [`zero_grad` the model params ](https://github.com/Lightning-AI/lightning/blob/343f80436d3233bac155e77b92891f6281ff90d1/src/lightning/pytorch/loops/evaluation_loop.py#L244)before entering the validation loop. 

In normal use cases, this `zero_grad` call is not noticeable. NeMo's MLPerf team reported however that

> touching the grads there is causing some of the all gathers we are doing  before the validation_step to get exposed and not allowing it to be overlapped with the compute of validation_step. We want the all_gathers of all the buckets other than first to be overlapped with the val_step.

**Option 1:** Remove the explicit `zero_grad` call as suggested by NeMo.
If this `zero_grad` call were to be removed, the implication is that grads will be available during validation (i.e. `all(p.grad is not None for p in self.parameters())` and potentially use more memory than expected. 

**Option 2:** Change the optimization step
An alternative would be to change the order to 1) `backward()` 3) `optimizer.step()` 3) `zero_grad()`. Then, an explicit zero-ing at the beginning of validation wouldn't be necessary. But this would potentially be a breaking change for some users who rely on this order and have overridden hooks accordingly.

**Option 3:** Make it possible to override the `zero_grad` call (turning it off) in the validation loop. For example, in the form of a hook `LightningModule.on_evaluation_zero_grad()` similar to our `on_validation_model_eval()` hooks.

cc @borda

**Comment by awaelchli** [comment]
Right now, the hook (Option 3) seems to be the safest moving forward. Please let me know if you have any other suggestions @carmocca @justusschock

**Comment by carmocca** [comment]
Option 2 is by design: https://github.com/Lightning-AI/lightning/pull/6147

Option 3: should be possible today by overriding `LigthningModule.optimizer_zero_grad` and checking whether we are validating.

Option 1 makes think: are we currently losing grads when validation runs before gradient accumulation is finished?

**Comment by awaelchli** [comment]
Yes agree, the optimization step probably can't be changed.

Option 3: should be possible today by overriding LigthningModule.optimizer_zero_grad and checking whether we are validating.

Note that the line in question is calling `model.zero_grad()`. So one could override this one yes. 

> Option 1 makes think: are we currently losing grads when validation runs before gradient accumulation is finished?

Yes, I tested it yesterday.

**Comment by erhoo82** [comment]
Regarding the memory saving with `zero_grad()`, it doesn't free (de-alloc) memory but keeps it as cached memory so we use the buffers for activation in the following validation step? Trying to understand the memory-saving mechanism here.

**Comment by erhoo82** [comment]
> One possibility would be to conditionally only zero_grad when the accumulation phase is over, and otherwise leave it.
This makes sense to me.

Option 3 is fine as long as we can override it.
